File: src/main/resources/public/ts/services/courseTag.service.ts
import {ng, Service} from "entcore";
import {CourseTag} from "../model/courseTag";
import http, {AxiosResponse} from 'axios';

export interface ICourseTagService {
    getCourseTags(structureId: string): Promise<Array<CourseTag>>;

    createCourseTag(structureId: string, courseTag: CourseTag): Promise<AxiosResponse>;

    updateCourseTagHidden(structureId: string, tagId: number, isHidden: boolean): Promise<AxiosResponse>;

    deleteCourseTag(structureId: string, tagId: number): Promise<AxiosResponse>;

    updateCourseTag(courseTag: CourseTag): Promise<AxiosResponse>;
}

export const courseTagService: ICourseTagService = {
    getCourseTags: async (structureId: string): Promise<Array<CourseTag>> => {
        return http.get(`/edt/structures/${structureId}/course/tags`)
            .then((res: AxiosResponse): Array<CourseTag> => {return res.data; });
    },

    createCourseTag: async (structureId: string, courseTag: CourseTag): Promise<AxiosResponse> => {
        return http.post(`/edt/structures/${structureId}/course/tag`, courseTag);
    },

    updateCourseTagHidden: async (structureId: string, tagId: number, isHidden: boolean): Promise<AxiosResponse> => {
        return http.put(`/edt/structures/${structureId}/course/tag/${tagId}/hidden`, {isHidden: isHidden});
    },

    deleteCourseTag: async (structureId: string, tagId: number): Promise<AxiosResponse> => {
        return http.delete(`/edt/structures/${structureId}/course/tag/${tagId}`);
    },

    updateCourseTag: async (courseTag: CourseTag): Promise<AxiosResponse> => {
        return http.put(`/edt/course/tag`, courseTag);
    }

};

export const CourseTagService: Service = ng.service('CourseTagService',
    (): ICourseTagService => courseTagService);

File: src/main/java/fr/cgi/edt/services/CourseTagService.java
package fr.cgi.edt.services;


import fr.cgi.edt.models.CourseTag;
import io.vertx.core.Future;
import io.vertx.core.json.JsonObject;

import java.util.List;

public interface CourseTagService {

    /**
     * Get all course tags from structure
     *
     * @param structureId Structure identifier
     */
    Future<List<CourseTag>> getCourseTags(String structureId);

    /**
     * Create a course tag for the structure
     * @param structureId       structure identifier
     * @param courseTagBody     course tag object
     */
    Future<JsonObject> createCourseTag(String structureId, JsonObject courseTagBody);


    /**
     * Delete a course tag from id
     * @param structureId       structure identifier
     * @param id                course tag identifier
     */
    Future<JsonObject> deleteCourseTag(String structureId, Number id);

    /**
     * Update a course tag
     * @param courseTagBody course tag object
     */
    Future<JsonObject> updateCourseTag(JsonObject courseTagBody);

    /**
     * Update a course tag priority
     * @param structureId      structure identifier
     * @param id               course tag identifier
     * @param isHidden         hidden state
     */
    Future<JsonObject> updateCourseTagHidden(String structureId, Number id, boolean isHidden);

}
